// @tact-lang/compiler/dist/config/features.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.featureEnable = exports.enabledInterfacesGetter = exports.enabledIpfsAbiGetter = exports.enabledExternals = exports.enabledMasterchain = exports.enabledDebug = exports.enabledInline = void 0;
const context_1 = require("../context");
const featureStore = (0, context_1.createContextStore)();
function enabledInline(ctx) {
    return featureEnabled(ctx, "inline");
}
exports.enabledInline = enabledInline;
function enabledDebug(ctx) {
    return featureEnabled(ctx, "debug");
}
exports.enabledDebug = enabledDebug;
function enabledMasterchain(ctx) {
    return featureEnabled(ctx, "masterchain");
}
exports.enabledMasterchain = enabledMasterchain;
function enabledExternals(ctx) {
    return featureEnabled(ctx, "external");
}
exports.enabledExternals = enabledExternals;
function enabledIpfsAbiGetter(ctx) {
    return featureEnabled(ctx, "ipfsAbiGetter");
}
exports.enabledIpfsAbiGetter = enabledIpfsAbiGetter;
function enabledInterfacesGetter(ctx) {
    return featureEnabled(ctx, "interfacesGetter");
}
exports.enabledInterfacesGetter = enabledInterfacesGetter;
function featureEnabled(ctx, key) {
    return featureStore.get(ctx, key) === true;
}
function featureEnable(ctx, key) {
    return featureStore.set(ctx, key, true);
}
exports.featureEnable = featureEnable;
