// @tact-lang/compiler/dist/optimizer/associative.js
"use strict";
// This module includes rules involving associative rewrites of expressions
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssociativeRule3 = exports.AssociativeRule2 = exports.AssociativeRule1 = void 0;
const constEval_1 = require("../constEval");
const ast_1 = require("../grammar/ast");
const types_1 = require("./types");
const util_1 = require("./util");
class AssociativeRewriteRule extends types_1.Rule {
    // An entry (op, S) in the map means "operator op associates with all operators in set S",
    // mathematically: all op2 \in S. (a op b) op2 c = a op (b op2 c)
    associativeOps;
    // This set contains all operators that commute.
    // Mathematically:
    // all op \in commutativeOps. a op b = b op a
    commutativeOps;
    constructor() {
        super();
        // + associates with these on the right:
        // i.e., all op \in additiveAssoc. (a + b) op c = a + (b op c)
        const additiveAssoc = new Set(["+", "-"]);
        // * associates with these on the right:
        const multiplicativeAssoc = new Set([
            "*",
            "<<",
        ]);
        // Division / does not associate with any on the right
        // Modulus % does not associate with any on the right
        // TODO: shifts, bitwise integer operators, boolean operators
        this.associativeOps = new Map([
            ["+", additiveAssoc],
            ["*", multiplicativeAssoc],
        ]);
        this.commutativeOps = new Set(["+", "*", "!=", "==", "&&", "||"]);
    }
    areAssociative(op1, op2) {
        if (this.associativeOps.has(op1)) {
            const rightOperators = this.associativeOps.get(op1);
            return rightOperators.has(op2);
        }
        else {
            return false;
        }
    }
    isCommutative(op) {
        return this.commutativeOps.has(op);
    }
}
class AllowableOpRule extends AssociativeRewriteRule {
    allowedOps;
    constructor() {
        super();
        this.allowedOps = new Set();
        // Recall that integer operators +,-,*,/,% are not safe with this rule, because
        // there is a risk that they will not preserve overflows in the unknown operands.
        //["&&", "||"], // TODO: check bitwise integer operators
    }
    isAllowedOp(op) {
        return this.allowedOps.has(op);
    }
    areAllowedOps(op) {
        return op.reduce((prev, curr) => prev && this.allowedOps.has(curr), true);
    }
}
// This rule will be removed in a future refactoring, since
// no operator can use it due to the safety conditions.
// At first I thought that boolean expressions could use them
// but I found out they cannot.
class AssociativeRule1 extends AllowableOpRule {
    applyRule(ast, optimizer) {
        if ((0, util_1.checkIsBinaryOpNode)(ast)) {
            const topLevelNode = ast;
            if ((0, util_1.checkIsBinaryOp_With_RightValue)(topLevelNode.left) &&
                (0, util_1.checkIsBinaryOp_With_RightValue)(topLevelNode.right)) {
                // The tree has this form:
                // (x1 op1 c1) op (x2 op2 c2)
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = leftTree.left;
                const c1 = leftTree.right;
                const op1 = leftTree.op;
                const x2 = rightTree.left;
                const c2 = rightTree.right;
                const op2 = rightTree.op;
                const op = topLevelNode.op;
                // Check that:
                // the operators are allowed
                // op1 and op associate
                // op and op2 associate
                // op commutes
                if (this.areAllowedOps([op1, op, op2]) &&
                    this.areAssociative(op1, op) &&
                    this.areAssociative(op, op2) &&
                    this.isCommutative(op)) {
                    // Agglutinate the constants and compute their final value
                    try {
                        // If an error occurs, we abandon the simplification
                        const val = (0, constEval_1.evalBinaryOp)(op2, (0, util_1.extractValue)(c1), (0, util_1.extractValue)(c2));
                        // The final expression is
                        // (x1 op1 x2) op val
                        // Because we are joining x1 and x2,
                        // there is further opportunity of simplification,
                        // So, we ask the evaluator to apply all the rules in the subtree.
                        const newLeft = optimizer.applyRules((0, util_1.makeBinaryExpression)(op1, x1, x2));
                        const newRight = (0, util_1.makeValueExpression)(val);
                        return (0, util_1.makeBinaryExpression)(op, newLeft, newRight);
                    }
                    catch (e) {
                        // Do nothing: will exit rule without modifying tree
                    }
                }
            }
            else if ((0, util_1.checkIsBinaryOp_With_RightValue)(topLevelNode.left) &&
                (0, util_1.checkIsBinaryOp_With_LeftValue)(topLevelNode.right)) {
                // The tree has this form:
                // (x1 op1 c1) op (c2 op2 x2)
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = leftTree.left;
                const c1 = leftTree.right;
                const op1 = leftTree.op;
                const x2 = rightTree.right;
                const c2 = rightTree.left;
                const op2 = rightTree.op;
                const op = topLevelNode.op;
                // Check that:
                // the operators are allowed
                // op1 and op associate
                // op and op2 associate
                if (this.areAllowedOps([op1, op, op2]) &&
                    this.areAssociative(op1, op) &&
                    this.areAssociative(op, op2)) {
                    // Agglutinate the constants and compute their final value
                    try {
                        // If an error occurs, we abandon the simplification
                        const val = (0, constEval_1.evalBinaryOp)(op, (0, util_1.extractValue)(c1), (0, util_1.extractValue)(c2));
                        // The current expression could be either
                        // x1 op1 (val op2 x2) or
                        // (x1 op1 val) op2 x2  <--- we choose this form.
                        // Other rules will attempt to extract the constant outside the expression.
                        // Because we are joining x1 and val,
                        // there is further opportunity of simplification,
                        // So, we ask the evaluator to apply all the rules in the subtree.
                        const newValNode = (0, util_1.makeValueExpression)(val);
                        const newLeft = optimizer.applyRules((0, util_1.makeBinaryExpression)(op1, x1, newValNode));
                        return (0, util_1.makeBinaryExpression)(op2, newLeft, x2);
                    }
                    catch (e) {
                        // Do nothing: will exit rule without modifying tree
                    }
                }
            }
            else if ((0, util_1.checkIsBinaryOp_With_LeftValue)(topLevelNode.left) &&
                (0, util_1.checkIsBinaryOp_With_RightValue)(topLevelNode.right)) {
                // The tree has this form:
                // (c1 op1 x1) op (x2 op2 c2)
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = leftTree.right;
                const c1 = leftTree.left;
                const op1 = leftTree.op;
                const x2 = rightTree.left;
                const c2 = rightTree.right;
                const op2 = rightTree.op;
                const op = topLevelNode.op;
                // Check that:
                // the operators are allowed
                // op and op1 associate
                // op2 and op associate
                // op commutes
                if (this.areAllowedOps([op1, op, op2]) &&
                    this.areAssociative(op, op1) &&
                    this.areAssociative(op2, op) &&
                    this.isCommutative(op)) {
                    // Agglutinate the constants and compute their final value
                    try {
                        // If an error occurs, we abandon the simplification
                        const val = (0, constEval_1.evalBinaryOp)(op, (0, util_1.extractValue)(c2), (0, util_1.extractValue)(c1));
                        // The current expression could be either
                        // x2 op2 (val op1 x1) or
                        // (x2 op2 val) op1 x1  <--- we choose this form.
                        // Other rules will attempt to extract the constant outside the expression.
                        // Because we are joining x2 and val,
                        // there is further opportunity of simplification,
                        // So, we ask the evaluator to apply all the rules in the subtree.
                        const newValNode = (0, util_1.makeValueExpression)(val);
                        const newLeft = optimizer.applyRules((0, util_1.makeBinaryExpression)(op2, x2, newValNode));
                        return (0, util_1.makeBinaryExpression)(op1, newLeft, x1);
                    }
                    catch (e) {
                        // Do nothing: will exit rule without modifying tree
                    }
                }
            }
            else if ((0, util_1.checkIsBinaryOp_With_LeftValue)(topLevelNode.left) &&
                (0, util_1.checkIsBinaryOp_With_LeftValue)(topLevelNode.right)) {
                // The tree has this form:
                // (c1 op1 x1) op (c2 op2 x2)
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = leftTree.right;
                const c1 = leftTree.left;
                const op1 = leftTree.op;
                const x2 = rightTree.right;
                const c2 = rightTree.left;
                const op2 = rightTree.op;
                const op = topLevelNode.op;
                // Check that:
                // the operators are allowed
                // op1 and op associate
                // op and op2 associate
                // op commutes
                if (this.areAllowedOps([op1, op, op2]) &&
                    this.areAssociative(op1, op) &&
                    this.areAssociative(op, op2) &&
                    this.isCommutative(op)) {
                    // Agglutinate the constants and compute their final value
                    try {
                        // If an error occurs, we abandon the simplification
                        const val = (0, constEval_1.evalBinaryOp)(op1, (0, util_1.extractValue)(c1), (0, util_1.extractValue)(c2));
                        // The final expression is
                        // val op (x1 op2 x2)
                        // Because we are joining x1 and x2,
                        // there is further opportunity of simplification,
                        // So, we ask the evaluator to apply all the rules in the subtree.
                        const newRight = optimizer.applyRules((0, util_1.makeBinaryExpression)(op2, x1, x2));
                        const newLeft = (0, util_1.makeValueExpression)(val);
                        return (0, util_1.makeBinaryExpression)(op, newLeft, newRight);
                    }
                    catch (e) {
                        // Do nothing: will exit rule without modifying tree
                    }
                }
            }
        }
        // If execution reaches here, it means that the rule could not be applied fully
        // so, we return the original tree
        return ast;
    }
}
exports.AssociativeRule1 = AssociativeRule1;
// This rule will be removed in a future refactoring, since
// no operator can use it due to the safety conditions.
// At first I thought that boolean expressions could use them
// but I found out they cannot.
class AssociativeRule2 extends AllowableOpRule {
    applyRule(ast, optimizer) {
        if ((0, util_1.checkIsBinaryOpNode)(ast)) {
            const topLevelNode = ast;
            if ((0, util_1.checkIsBinaryOp_With_RightValue)(topLevelNode.left) &&
                !(0, ast_1.isValue)(topLevelNode.right)) {
                // The tree has this form:
                // (x1 op1 c1) op x2
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = leftTree.left;
                const c1 = leftTree.right;
                const op1 = leftTree.op;
                const x2 = rightTree;
                const op = topLevelNode.op;
                // Check that:
                // the operators are allowed
                // op1 and op associate
                // op commutes
                if (this.areAllowedOps([op1, op]) &&
                    this.areAssociative(op1, op) &&
                    this.isCommutative(op)) {
                    // The final expression is
                    // (x1 op1 x2) op c1
                    // Because we are joining x1 and x2,
                    // there is further opportunity of simplification,
                    // So, we ask the evaluator to apply all the rules in the subtree.
                    const newLeft = optimizer.applyRules((0, util_1.makeBinaryExpression)(op1, x1, x2));
                    return (0, util_1.makeBinaryExpression)(op, newLeft, c1);
                }
            }
            else if ((0, util_1.checkIsBinaryOp_With_LeftValue)(topLevelNode.left) &&
                !(0, ast_1.isValue)(topLevelNode.right)) {
                // The tree has this form:
                // (c1 op1 x1) op x2
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = leftTree.right;
                const c1 = leftTree.left;
                const op1 = leftTree.op;
                const x2 = rightTree;
                const op = topLevelNode.op;
                // Check that:
                // the operators are allowed
                // op1 and op associate
                if (this.areAllowedOps([op1, op]) &&
                    this.areAssociative(op1, op)) {
                    // The final expression is
                    // c1 op1 (x1 op x2)
                    // Because we are joining x1 and x2,
                    // there is further opportunity of simplification,
                    // So, we ask the evaluator to apply all the rules in the subtree.
                    const newRight = optimizer.applyRules((0, util_1.makeBinaryExpression)(op, x1, x2));
                    return (0, util_1.makeBinaryExpression)(op1, c1, newRight);
                }
            }
            else if (!(0, ast_1.isValue)(topLevelNode.left) &&
                (0, util_1.checkIsBinaryOp_With_RightValue)(topLevelNode.right)) {
                // The tree has this form:
                // x2 op (x1 op1 c1)
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = rightTree.left;
                const c1 = rightTree.right;
                const op1 = rightTree.op;
                const x2 = leftTree;
                const op = topLevelNode.op;
                // Check that:
                // the operators are allowed
                // op and op1 associate
                if (this.areAllowedOps([op, op1]) &&
                    this.areAssociative(op, op1)) {
                    // The final expression is
                    // (x2 op x1) op1 c1
                    // Because we are joining x1 and x2,
                    // there is further opportunity of simplification,
                    // So, we ask the evaluator to apply all the rules in the subtree.
                    const newLeft = optimizer.applyRules((0, util_1.makeBinaryExpression)(op, x2, x1));
                    return (0, util_1.makeBinaryExpression)(op1, newLeft, c1);
                }
            }
            else if (!(0, ast_1.isValue)(topLevelNode.left) &&
                (0, util_1.checkIsBinaryOp_With_LeftValue)(topLevelNode.right)) {
                // The tree has this form:
                // x2 op (c1 op1 x1)
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = rightTree.right;
                const c1 = rightTree.left;
                const op1 = rightTree.op;
                const x2 = leftTree;
                const op = topLevelNode.op;
                // Check that:
                // the operators are allowed
                // op and op1 associate
                // op is commutative
                if (this.areAllowedOps([op, op1]) &&
                    this.areAssociative(op, op1) &&
                    this.isCommutative(op)) {
                    // The final expression is
                    // c1 op (x2 op1 x1)
                    // Because we are joining x1 and x2,
                    // there is further opportunity of simplification,
                    // So, we ask the evaluator to apply all the rules in the subtree.
                    const newRight = optimizer.applyRules((0, util_1.makeBinaryExpression)(op1, x2, x1));
                    return (0, util_1.makeBinaryExpression)(op, c1, newRight);
                }
            }
        }
        // If execution reaches here, it means that the rule could not be applied fully
        // so, we return the original tree
        return ast;
    }
}
exports.AssociativeRule2 = AssociativeRule2;
class AssociativeRule3 extends types_1.Rule {
    leftAssocTransforms;
    rightAssocTransforms;
    rightCommuteTransforms;
    leftCommuteTransforms;
    // Safety conditions that repeat a lot.
    // Safety condition:
    // c1 < center ==> val_ - c1 <= 0
    // c1 > center ==> val_ - c1 >= 0
    standardAdditiveCondition(c1, val_, center) {
        if (c1 === center) {
            return true;
        }
        else if (c1 < center) {
            return val_ - c1 <= 0n;
        }
        else {
            return val_ - c1 >= 0n;
        }
    }
    // Safety condition:
    // c1 < 0 ==> val_ - c1 <= -1
    // c1 > 0 ==> val_ - c1 >= -1
    shiftedAdditiveCondition(c1, val_) {
        if (c1 === 0n) {
            return true;
        }
        else if (c1 < 0n) {
            return val_ - c1 <= -1n;
        }
        else {
            return val_ - c1 >= -1n;
        }
    }
    // Safety condition:
    // c1 < center ==> val_ + c1 >= -1
    // c1 > center ==> val_ + c1 <= -1
    oppositeAdditiveCondition(c1, val_, center) {
        if (c1 === center) {
            return true;
        }
        else if (c1 < center) {
            return val_ + c1 >= -1n;
        }
        else {
            return val_ + c1 <= -1n;
        }
    }
    // Safety condition:
    // c1 != 0 && sign(c1) == sign(val_) ==> abs(c1) <= abs(val_)
    // c1 != 0 && sign(c1) != sign(val_) ==> abs(c1) < abs(val_)
    // c1 != 0 ==> val_ != 0
    standardMultiplicativeCondition(c1, val_) {
        if (c1 === 0n) {
            return true;
        }
        // At this point, c1 != 0
        // hence, val_ must be non-zero as
        // required by the safety condition
        if (val_ === 0n) {
            return false;
        }
        // At this point, both c1 and val_ are non-zero
        if ((0, util_1.sign)(c1) === (0, util_1.sign)(val_)) {
            return (0, util_1.abs)(c1) <= (0, util_1.abs)(val_);
        }
        else {
            return (0, util_1.abs)(c1) < (0, util_1.abs)(val_);
        }
    }
    constructor() {
        super();
        // First, we consider expressions of the form: (x1 op1 c1) op c2.
        // The following maps correspond to the transformation: x1 op1_ (c1_ op_ c2_)
        // for each pair of operators op1, op.
        // Here, we will denote c1_ op_ c2_ as val_.
        // op1 = +
        const plusLeftAssocOperators = new Map([
            [
                "+",
                // original expression: (x1 + c1) + c2
                (x1, c1, c2) => {
                    // final expression: x1 + (c1 + c2)
                    const val_ = (0, constEval_1.evalBinaryOp)("+", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("+", x1, val_node),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
            [
                "-",
                // original expression: (x1 + c1) - c2
                (x1, c1, c2) => {
                    // final expression: x1 + (c1 - c2)
                    const val_ = (0, constEval_1.evalBinaryOp)("-", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("+", x1, val_node),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
        ]);
        // op1 = -
        const minusLeftAssocOperators = new Map([
            [
                "+",
                // original expression: (x1 - c1) + c2
                (x1, c1, c2) => {
                    // final expression x1 - (c1 - c2)
                    const val_ = (0, constEval_1.evalBinaryOp)("-", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("-", x1, val_node),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
            [
                "-",
                // original expression: (x1 - c1) - c2
                (x1, c1, c2) => {
                    // final expression x1 - (c1 + c2)
                    const val_ = (0, constEval_1.evalBinaryOp)("+", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("-", x1, val_node),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
        ]);
        // op1 = *
        const multiplyLeftAssocOperators = new Map([
            [
                "*",
                // original expression: (x1 * c1) * c2
                (x1, c1, c2) => {
                    // final expression x1 * (c1 * c2)
                    const val_ = (0, constEval_1.evalBinaryOp)("*", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("*", x1, val_node),
                        safetyCondition: this.standardMultiplicativeCondition(c1, val_),
                    };
                },
            ],
        ]);
        // op1 = &&
        const andLeftAssocOperators = new Map([
            [
                "&&",
                // original expression: (x1 && c1) && c2
                (x1, c1, c2) => {
                    // final expression x1 && (c1 && c2)
                    const val_ = (0, constEval_1.evalBinaryOp)("&&", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("&&", x1, val_node),
                        safetyCondition: true,
                    };
                },
            ],
        ]);
        // op1 = ||
        const orLeftAssocOperators = new Map([
            [
                "||",
                // original expression: (x1 || c1) || c2
                (x1, c1, c2) => {
                    // final expression x1 || (c1 || c2)
                    const val_ = (0, constEval_1.evalBinaryOp)("||", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("||", x1, val_node),
                        safetyCondition: true,
                    };
                },
            ],
        ]);
        this.leftAssocTransforms = new Map([
            ["+", plusLeftAssocOperators],
            ["-", minusLeftAssocOperators],
            ["*", multiplyLeftAssocOperators],
            ["&&", andLeftAssocOperators],
            ["||", orLeftAssocOperators],
        ]);
        // Now consider expressions of the form: c2 op (c1 op1 x1).
        // The following maps correspond to the transformation: (c2_ op_ c1_) op1_ x1
        // for each pair of operators op1, op.
        // Here, we will denote c2_ op_ c1_ as val_.
        // op = +
        const plusRightAssocOperators = new Map([
            [
                "+",
                // original expression: c2 + (c1 + x1)
                (x1, c1, c2) => {
                    // final expression (c2 + c1) + x1
                    const val_ = (0, constEval_1.evalBinaryOp)("+", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("+", val_node, x1),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
            [
                "-",
                // original expression: c2 + (c1 - x1)
                (x1, c1, c2) => {
                    // final expression (c2 + c1) - x1
                    const val_ = (0, constEval_1.evalBinaryOp)("+", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("-", val_node, x1),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, -1n),
                    };
                },
            ],
        ]);
        // op = -
        const minusRightAssocOperators = new Map([
            [
                "+",
                // original expression: c2 - (c1 + x1)
                (x1, c1, c2) => {
                    // final expression (c2 - c1) - x1
                    const val_ = (0, constEval_1.evalBinaryOp)("-", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("-", val_node, x1),
                        safetyCondition: this.oppositeAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
            [
                "-",
                // original expression: c2 - (c1 - x1)
                (x1, c1, c2) => {
                    // final expression (c2 - c1) + x1
                    const val_ = (0, constEval_1.evalBinaryOp)("-", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("+", val_node, x1),
                        safetyCondition: this.oppositeAdditiveCondition(c1, val_, -1n),
                    };
                },
            ],
        ]);
        // op = *
        const multiplyRightAssocOperators = new Map([
            [
                "*",
                // original expression: c2 * (c1 * x1)
                (x1, c1, c2) => {
                    // final expression (c2 * c1) * x1
                    const val_ = (0, constEval_1.evalBinaryOp)("*", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("*", val_node, x1),
                        safetyCondition: this.standardMultiplicativeCondition(c1, val_),
                    };
                },
            ],
        ]);
        // op = &&
        const andRightAssocOperators = new Map([
            [
                "&&",
                // original expression: c2 && (c1 && x1)
                (x1, c1, c2) => {
                    // final expression (c2 && c1) && x1
                    const val_ = (0, constEval_1.evalBinaryOp)("&&", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("&&", val_node, x1),
                        safetyCondition: true,
                    };
                },
            ],
        ]);
        // op = ||
        const orRightAssocOperators = new Map([
            [
                "||",
                // original expression: c2 || (c1 || x1)
                (x1, c1, c2) => {
                    // final expression (c2 || c1) || x1
                    const val_ = (0, constEval_1.evalBinaryOp)("||", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("||", val_node, x1),
                        safetyCondition: true,
                    };
                },
            ],
        ]);
        this.rightAssocTransforms = new Map([
            ["+", plusRightAssocOperators],
            ["-", minusRightAssocOperators],
            ["*", multiplyRightAssocOperators],
            ["&&", andRightAssocOperators],
            ["||", orRightAssocOperators],
        ]);
        // Now consider expressions of the form: c2 op (x1 op1 c1).
        // The following maps correspond to the transformation: x1 op1_ (c2_ op_ c1_)_
        // for each pair of operators op1, op.
        // Here, we will denote c2_ op_ c1_ as val_.
        // op = +
        const plusRightCommuteOperators = new Map([
            [
                "+",
                // original expression: c2 + (x1 + c1)
                (x1, c1, c2) => {
                    // final expression x1 + (c2 + c1)
                    const val_ = (0, constEval_1.evalBinaryOp)("+", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("+", x1, val_node),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
            [
                "-",
                // original expression: c2 + (x1 - c1)
                (x1, c1, c2) => {
                    // final expression x1 - (c1 - c2)
                    const val_ = (0, constEval_1.evalBinaryOp)("-", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("-", x1, val_node),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
        ]);
        // op = -
        const minusRightCommuteOperators = new Map([
            [
                "+",
                // original expression: c2 - (x1 + c1)
                (x1, c1, c2) => {
                    // final expression (c2 - c1) - x1
                    const val_ = (0, constEval_1.evalBinaryOp)("-", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("-", val_node, x1),
                        safetyCondition: this.oppositeAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
            [
                "-",
                // original expression: c2 - (x1 - c1)
                (x1, c1, c2) => {
                    // final expression (c2 + c1) - x1
                    const val_ = (0, constEval_1.evalBinaryOp)("+", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("-", val_node, x1),
                        safetyCondition: this.shiftedAdditiveCondition(c1, val_),
                    };
                },
            ],
        ]);
        // op = *
        const multiplyRightCommuteOperators = new Map([
            [
                "*",
                // original expression: c2 * (x1 * c1)
                (x1, c1, c2) => {
                    // Final expression x1 * (c2 * c1)
                    const val_ = (0, constEval_1.evalBinaryOp)("*", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("*", x1, val_node),
                        safetyCondition: this.standardMultiplicativeCondition(c1, val_),
                    };
                },
            ],
        ]);
        // op = &&
        const andRightCommuteOperators = new Map([
            [
                "&&",
                // original expression: c2 && (x1 && c1)
                (x1, c1, c2) => {
                    const val_ = (0, constEval_1.evalBinaryOp)("&&", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    let final_expr;
                    if (c2 === true) {
                        // Final expression x1 && (c2 && c1)
                        final_expr = (0, util_1.makeBinaryExpression)("&&", x1, val_node);
                    }
                    else {
                        // Final expression (c2 && c1) && x1
                        // Note that by the safety condition,
                        // at this point c1 = true.
                        final_expr = (0, util_1.makeBinaryExpression)("&&", val_node, x1);
                    }
                    return {
                        simplifiedExpression: final_expr,
                        safetyCondition: c1 === true || c2 === true,
                    };
                },
            ],
        ]);
        // op = ||
        const orRightCommuteOperators = new Map([
            [
                "||",
                // original expression: c2 || (x1 || c1)
                (x1, c1, c2) => {
                    const val_ = (0, constEval_1.evalBinaryOp)("||", c2, c1);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    let final_expr;
                    if (c2 === false) {
                        // Final expression x1 || (c2 || c1)
                        final_expr = (0, util_1.makeBinaryExpression)("||", x1, val_node);
                    }
                    else {
                        // Final expression (c2 || c1) || x1
                        // Note that by the safety condition,
                        // at this point c1 = false.
                        final_expr = (0, util_1.makeBinaryExpression)("||", val_node, x1);
                    }
                    return {
                        simplifiedExpression: final_expr,
                        safetyCondition: c1 === false || c2 === false,
                    };
                },
            ],
        ]);
        this.rightCommuteTransforms = new Map([
            ["+", plusRightCommuteOperators],
            ["-", minusRightCommuteOperators],
            ["*", multiplyRightCommuteOperators],
            ["&&", andRightCommuteOperators],
            ["||", orRightCommuteOperators],
        ]);
        // Now consider expressions of the form: (c1 op1 x1) op c2.
        // The following maps correspond to the transformation: x1 op1_ (c1_ op_ c2_)
        // for each pair of operators op1, op.
        // op1 = +
        const plusLeftCommuteOperators = new Map([
            [
                "+",
                // original expression: (c1 + x1) + c2
                (x1, c1, c2) => {
                    // Final expression (c1 + c2) + x1
                    const val_ = (0, constEval_1.evalBinaryOp)("+", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("+", val_node, x1),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
            [
                "-",
                // original expression: (c1 + x1) - c2
                (x1, c1, c2) => {
                    // Final expression (c1 - c2) + x1
                    const val_ = (0, constEval_1.evalBinaryOp)("-", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("+", val_node, x1),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, 0n),
                    };
                },
            ],
        ]);
        // op1 = -
        const minusLeftCommuteOperators = new Map([
            [
                "+",
                // original expression: (c1 - x1) + c2
                (x1, c1, c2) => {
                    // Final expression (c1 + c2) - x1
                    const val_ = (0, constEval_1.evalBinaryOp)("+", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("-", val_node, x1),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, -1n),
                    };
                },
            ],
            [
                "-",
                // original expression: (c1 - x1) - c2
                (x1, c1, c2) => {
                    // Final expression (c1 - c2) - x1
                    const val_ = (0, constEval_1.evalBinaryOp)("-", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("-", val_node, x1),
                        safetyCondition: this.standardAdditiveCondition(c1, val_, -1n),
                    };
                },
            ],
        ]);
        // op1 = *
        const multiplyLeftCommuteOperators = new Map([
            [
                "*",
                // original expression: (c1 * x1) * c2
                (x1, c1, c2) => {
                    // Final expression (c1 * c2) * x1
                    const val_ = (0, constEval_1.evalBinaryOp)("*", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    return {
                        simplifiedExpression: (0, util_1.makeBinaryExpression)("*", val_node, x1),
                        safetyCondition: this.standardMultiplicativeCondition(c1, val_),
                    };
                },
            ],
        ]);
        // op1 = &&
        const andLeftCommuteOperators = new Map([
            [
                "&&",
                // original expression: (c1 && x1) && c2
                (x1, c1, c2) => {
                    const val_ = (0, constEval_1.evalBinaryOp)("&&", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    let final_expr;
                    if (c2 === true) {
                        // Final expression (c1 && c2) && x1
                        final_expr = (0, util_1.makeBinaryExpression)("&&", val_node, x1);
                    }
                    else {
                        // Final expression x1 && (c1 && c2)
                        // Note that by the safety condition,
                        // at this point c1 = true.
                        final_expr = (0, util_1.makeBinaryExpression)("&&", x1, val_node);
                    }
                    return {
                        simplifiedExpression: final_expr,
                        safetyCondition: c1 === true || c2 === true,
                    };
                },
            ],
        ]);
        // op1 = ||
        const orLeftCommuteOperators = new Map([
            [
                "||",
                // original expression: (c1 || x1) || c2
                (x1, c1, c2) => {
                    const val_ = (0, constEval_1.evalBinaryOp)("||", c1, c2);
                    const val_node = (0, util_1.makeValueExpression)(val_);
                    let final_expr;
                    if (c2 === false) {
                        // Final expression (c1 || c2) || x1
                        final_expr = (0, util_1.makeBinaryExpression)("||", val_node, x1);
                    }
                    else {
                        // Final expression x1 || (c1 || c2)
                        // Note that by the safety condition,
                        // at this point c1 = false.
                        final_expr = (0, util_1.makeBinaryExpression)("||", x1, val_node);
                    }
                    return {
                        simplifiedExpression: final_expr,
                        safetyCondition: c1 === false || c2 === false,
                    };
                },
            ],
        ]);
        this.leftCommuteTransforms = new Map([
            ["+", plusLeftCommuteOperators],
            ["-", minusLeftCommuteOperators],
            ["*", multiplyLeftCommuteOperators],
            ["&&", andLeftCommuteOperators],
            ["||", orLeftCommuteOperators],
        ]);
    }
    lookupTransform(keyOp1, keyOp2, transforms) {
        if (transforms.has(keyOp1)) {
            const intermediateMap = transforms.get(keyOp1);
            if (intermediateMap.has(keyOp2)) {
                return intermediateMap.get(keyOp2);
            }
        }
        return undefined;
    }
    getLeftAssociativityTransform(keyOp1, keyOp2) {
        return this.lookupTransform(keyOp1, keyOp2, this.leftAssocTransforms);
    }
    getRightAssociativityTransform(keyOp1, keyOp2) {
        return this.lookupTransform(keyOp1, keyOp2, this.rightAssocTransforms);
    }
    getLeftCommutativityTransform(keyOp1, keyOp2) {
        return this.lookupTransform(keyOp1, keyOp2, this.leftCommuteTransforms);
    }
    getRightCommutativityTransform(keyOp1, keyOp2) {
        return this.lookupTransform(keyOp1, keyOp2, this.rightCommuteTransforms);
    }
    applyRule(ast, optimizer) {
        if ((0, util_1.checkIsBinaryOpNode)(ast)) {
            const topLevelNode = ast;
            if ((0, util_1.checkIsBinaryOp_With_RightValue)(topLevelNode.left) &&
                (0, ast_1.isValue)(topLevelNode.right)) {
                // The tree has this form:
                // (x1 op1 c1) op c2
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = leftTree.left;
                const c1 = (0, util_1.extractValue)(leftTree.right);
                const op1 = leftTree.op;
                const c2 = (0, util_1.extractValue)(rightTree);
                const op = topLevelNode.op;
                try {
                    const data = this.getLeftAssociativityTransform(op1, op)(x1, c1, c2);
                    if (data.safetyCondition) {
                        // Since the tree is simpler now, there is further
                        // opportunity for simplification that was missed
                        // previously
                        return optimizer.applyRules(data.simplifiedExpression);
                    }
                }
                catch (e) {
                    // Do nothing: will exit rule without modifying tree
                }
            }
            else if ((0, util_1.checkIsBinaryOp_With_LeftValue)(topLevelNode.left) &&
                (0, ast_1.isValue)(topLevelNode.right)) {
                // The tree has this form:
                // (c1 op1 x1) op c2
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = leftTree.right;
                const c1 = (0, util_1.extractValue)(leftTree.left);
                const op1 = leftTree.op;
                const c2 = (0, util_1.extractValue)(rightTree);
                const op = topLevelNode.op;
                try {
                    const data = this.getLeftCommutativityTransform(op1, op)(x1, c1, c2);
                    if (data.safetyCondition) {
                        // Since the tree is simpler now, there is further
                        // opportunity for simplification that was missed
                        // previously
                        return optimizer.applyRules(data.simplifiedExpression);
                    }
                }
                catch (e) {
                    // Do nothing: will exit rule without modifying tree
                }
            }
            else if ((0, ast_1.isValue)(topLevelNode.left) &&
                (0, util_1.checkIsBinaryOp_With_RightValue)(topLevelNode.right)) {
                // The tree has this form:
                // c2 op (x1 op1 c1)
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = rightTree.left;
                const c1 = (0, util_1.extractValue)(rightTree.right);
                const op1 = rightTree.op;
                const c2 = (0, util_1.extractValue)(leftTree);
                const op = topLevelNode.op;
                try {
                    const data = this.getRightCommutativityTransform(op, op1)(x1, c1, c2);
                    if (data.safetyCondition) {
                        // Since the tree is simpler now, there is further
                        // opportunity for simplification that was missed
                        // previously
                        return optimizer.applyRules(data.simplifiedExpression);
                    }
                }
                catch (e) {
                    // Do nothing: will exit rule without modifying tree
                }
            }
            else if ((0, ast_1.isValue)(topLevelNode.left) &&
                (0, util_1.checkIsBinaryOp_With_LeftValue)(topLevelNode.right)) {
                // The tree has this form:
                // c2 op (c1 op1 x1)
                const leftTree = topLevelNode.left;
                const rightTree = topLevelNode.right;
                const x1 = rightTree.right;
                const c1 = (0, util_1.extractValue)(rightTree.left);
                const op1 = rightTree.op;
                const c2 = (0, util_1.extractValue)(leftTree);
                const op = topLevelNode.op;
                try {
                    const data = this.getRightAssociativityTransform(op, op1)(x1, c1, c2);
                    if (data.safetyCondition) {
                        // Since the tree is simpler now, there is further
                        // opportunity for simplification that was missed
                        // previously
                        return optimizer.applyRules(data.simplifiedExpression);
                    }
                }
                catch (e) {
                    // Do nothing: will exit rule without modifying tree
                }
            }
        }
        // If execution reaches here, it means that the rule could not be applied fully
        // so, we return the original tree
        return ast;
    }
}
exports.AssociativeRule3 = AssociativeRule3;
