// @tact-lang/compiler/dist/grammar/clone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cloneNode = void 0;
const ast_1 = require("./ast");
const errors_1 = require("../errors");
function cloneNode(src) {
    if (src.kind === "boolean") {
        return (0, ast_1.cloneAstNode)(src);
    }
    else if (src.kind === "id") {
        return (0, ast_1.cloneAstNode)(src);
    }
    else if (src.kind === "null") {
        return (0, ast_1.cloneAstNode)(src);
    }
    else if (src.kind === "number") {
        return (0, ast_1.cloneAstNode)(src);
    }
    else if (src.kind === "string") {
        return (0, ast_1.cloneAstNode)(src);
    }
    else if (src.kind === "statement_assign") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            path: cloneNode(src.path),
            expression: cloneNode(src.expression),
        });
    }
    else if (src.kind === "statement_augmentedassign") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            path: cloneNode(src.path),
            expression: cloneNode(src.expression),
        });
    }
    else if (src.kind === "statement_let") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            type: src.type ? (0, ast_1.cloneAstNode)(src.type) : null,
            expression: cloneNode(src.expression),
        });
    }
    else if (src.kind === "statement_condition") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            condition: cloneNode(src.condition),
            trueStatements: src.trueStatements.map(cloneNode),
            falseStatements: src.falseStatements
                ? src.falseStatements.map(cloneNode)
                : null,
            elseif: src.elseif ? cloneNode(src.elseif) : null,
        });
    }
    else if (src.kind === "struct_field_initializer") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            initializer: cloneNode(src.initializer),
        });
    }
    else if (src.kind === "statement_expression") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            expression: cloneNode(src.expression),
        });
    }
    else if (src.kind === "op_binary") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            left: cloneNode(src.left),
            right: cloneNode(src.right),
        });
    }
    else if (src.kind === "op_unary") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            operand: cloneNode(src.operand),
        });
    }
    else if (src.kind === "struct_instance") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            args: src.args.map(cloneNode),
        });
    }
    else if (src.kind === "method_call") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            self: cloneNode(src.self),
            args: src.args.map(cloneNode),
        });
    }
    else if (src.kind === "field_access") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            aggregate: cloneNode(src.aggregate),
        });
    }
    else if (src.kind === "static_call") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            args: src.args.map(cloneNode),
        });
    }
    else if (src.kind === "conditional") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            condition: cloneNode(src.condition),
            thenBranch: cloneNode(src.thenBranch),
            elseBranch: cloneNode(src.elseBranch),
        });
    }
    else if (src.kind === "statement_return") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            expression: src.expression ? cloneNode(src.expression) : null,
        });
    }
    else if (src.kind === "statement_repeat") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            iterations: cloneNode(src.iterations),
            statements: src.statements.map(cloneNode),
        });
    }
    else if (src.kind === "statement_until") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            condition: cloneNode(src.condition),
            statements: src.statements.map(cloneNode),
        });
    }
    else if (src.kind === "statement_while") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            condition: cloneNode(src.condition),
            statements: src.statements.map(cloneNode),
        });
    }
    else if (src.kind === "statement_try") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            statements: src.statements.map(cloneNode),
        });
    }
    else if (src.kind === "statement_try_catch") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            statements: src.statements.map(cloneNode),
            catchStatements: src.catchStatements.map(cloneNode),
        });
    }
    else if (src.kind === "statement_foreach") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            statements: src.statements.map(cloneNode),
        });
    }
    else if (src.kind === "function_def") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            return: src.return ? (0, ast_1.cloneAstNode)(src.return) : null,
            statements: src.statements.map(cloneNode),
            params: src.params.map(cloneNode),
        });
    }
    else if (src.kind === "function_decl") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            return: src.return ? (0, ast_1.cloneAstNode)(src.return) : null,
            params: src.params.map(cloneNode),
        });
    }
    else if (src.kind === "native_function_decl") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            return: src.return ? (0, ast_1.cloneAstNode)(src.return) : null,
            params: src.params.map(cloneNode),
        });
    }
    else if (src.kind === "receiver") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            statements: src.statements.map(cloneNode),
        });
    }
    else if (src.kind === "typed_parameter") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            type: (0, ast_1.cloneAstNode)(src.type),
        });
    }
    else if (src.kind === "init_of") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            args: src.args.map(cloneNode),
        });
    }
    else if (src.kind === "constant_def") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            type: (0, ast_1.cloneAstNode)(src.type),
            initializer: cloneNode(src.initializer),
        });
    }
    else if (src.kind === "constant_decl") {
        return (0, ast_1.cloneAstNode)({
            ...src,
            type: (0, ast_1.cloneAstNode)(src.type),
        });
    }
    (0, errors_1.throwInternalCompilerError)(`Not implemented for ${src.kind}`);
}
exports.cloneNode = cloneNode;
