// @ton/blueprint/dist/utils/selection.utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectFile = exports.selectOption = exports.findScripts = exports.findCompiles = void 0;
const path_1 = __importDefault(require("path"));
const promises_1 = __importDefault(require("fs/promises"));
const paths_1 = require("../paths");
const compile_1 = require("../compile/compile");
const findCompiles = async (directory) => {
    const dir = directory ?? await (0, compile_1.getCompilablesDirectory)();
    const files = await promises_1.default.readdir(dir);
    const compilables = files.filter((file) => file.endsWith(compile_1.COMPILE_END));
    return compilables.map((file) => ({
        path: path_1.default.join(dir, file),
        name: file.slice(0, file.length - compile_1.COMPILE_END.length),
    }));
};
exports.findCompiles = findCompiles;
const findScripts = async () => {
    const dirents = await promises_1.default.readdir(paths_1.SCRIPTS_DIR, { recursive: true, withFileTypes: true });
    const scripts = dirents.filter((dirent) => dirent.isFile() && dirent.name.endsWith('.ts'));
    return scripts
        .map((script) => ({
        name: path_1.default.join(script.path.slice(paths_1.SCRIPTS_DIR.length + 1), path_1.default.parse(script.name).name),
        path: path_1.default.join(script.path, script.name),
    }))
        .sort((a, b) => (a.name >= b.name ? 1 : -1));
};
exports.findScripts = findScripts;
async function selectOption(options, opts) {
    if (opts.hint) {
        const found = options.find((o) => o.value === opts.hint);
        if (found === undefined) {
            throw new Error(`Could not find option '${opts.hint}'`);
        }
        return found;
    }
    else {
        return await opts.ui.choose(opts.msg, options, (o) => o.name);
    }
}
exports.selectOption = selectOption;
async function selectFile(files, opts) {
    let selected;
    if (opts.hint) {
        const found = files.find((f) => f.name.toLowerCase() === opts.hint?.toLowerCase());
        if (found === undefined) {
            throw new Error(`Could not find file with name '${opts.hint}'`);
        }
        selected = found;
        opts.ui.write(`Using file: ${selected.name}`);
    }
    else {
        if (files.length === 1) {
            selected = files[0];
            opts.ui.write(`Using file: ${selected.name}`);
        }
        else {
            selected = await opts.ui.choose('Choose file to use', files, (f) => f.name);
        }
    }
    return {
        ...selected,
        module: opts.import !== false ? await (_a = selected.path, Promise.resolve().then(() => __importStar(require(_a)))) : undefined,
    };
}
exports.selectFile = selectFile;
