// @ton/blueprint/dist/paths.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TACT_ROOT_CONFIG = exports.BLUEPRINT_CONFIG = exports.TESTS_DIR = exports.CONTRACTS_DIR = exports.TEMP_DIR = exports.BUILD_DIR = exports.SCRIPTS_DIR = exports.WRAPPERS_DIR = exports.COMPILABLES_DIR = exports.BUILD = exports.TEMP = exports.SCRIPTS = exports.WRAPPERS = exports.COMPILABLES = exports.TESTS = exports.CONTRACTS = void 0;
const path_1 = __importDefault(require("path"));
exports.CONTRACTS = 'contracts';
exports.TESTS = 'tests';
exports.COMPILABLES = 'compilables';
exports.WRAPPERS = 'wrappers';
exports.SCRIPTS = 'scripts';
exports.TEMP = 'temp';
exports.BUILD = 'build';
exports.COMPILABLES_DIR = path_1.default.join(process.cwd(), exports.COMPILABLES);
exports.WRAPPERS_DIR = path_1.default.join(process.cwd(), exports.WRAPPERS);
exports.SCRIPTS_DIR = path_1.default.join(process.cwd(), exports.SCRIPTS);
exports.BUILD_DIR = path_1.default.join(process.cwd(), exports.BUILD);
exports.TEMP_DIR = path_1.default.join(process.cwd(), exports.TEMP);
exports.CONTRACTS_DIR = path_1.default.join(process.cwd(), exports.CONTRACTS);
exports.TESTS_DIR = path_1.default.join(process.cwd(), exports.TESTS);
exports.BLUEPRINT_CONFIG = path_1.default.join(process.cwd(), 'blueprint.config.ts');
exports.TACT_ROOT_CONFIG = path_1.default.join(process.cwd(), 'tact.config.json');
